### PR TITLE
GA4GH updates for e85

### PR DESCRIPTION
**Pull request by sarahhunt**
This adds 7 new GA4GH endpoints for features & variant annotation support. Existing endpoints have also been updated to return empty arrays when a page size of zero is requested rather than using the default page size. 
Requires: https://github.com/Ensembl/ensembl-rest_private/pull/7

### CODE REVIEW

File: lib/EnsEMBL/REST/Model/ga4gh/annotationSets.pm [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+use Moose;
+extends 'Catalyst::Model';
+use Catalyst::Exception;
+
+use Scalar::Util qw/weaken/;
+
+use EnsEMBL::REST::Model::ga4gh::ga4gh_utils;
+with 'Catalyst::Component::InstancePerContext';
+
+has 'context' => (is => 'ro');
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/ga4gh/annotationSets.pm]: Needs to be weakened or we get memory leaks.

File: lib/EnsEMBL/REST/Model/ga4gh/featureSets.pm [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+use Moose;
+extends 'Catalyst::Model';
+use Catalyst::Exception;
+
+use Scalar::Util qw/weaken/;
+
+use EnsEMBL::REST::Model::ga4gh::ga4gh_utils;
+with 'Catalyst::Component::InstancePerContext';
+
+has 'context' => (is => 'ro');
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/ga4gh/featureSets.pm]: Also needs weakening. weak_ref => 1,

File: lib/EnsEMBL/REST/Model/ga4gh/features.pm [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+
+}
+
+########### look up common values
+
+=head2 fetchSO
+      look up onology information for type
+      return GA4GH OntologyTerm
+=cut
+sub fetchSO{
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/ga4gh/features.pm]: Boldly assumes successful database fetch

File: lib/EnsEMBL/REST/Model/ga4gh/variantannotations.pm [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+  my @var_ann;
+  my $nextPageToken;
+  ## VCF is 1-based; GA4GH is 0-based & open-closed 
+  $data->{end}++;
+  $data->{pageToken} = $data->{start} unless defined $data->{pageToken}  && $data->{pageToken}  =~/\d+/;
+
+  ## exits here if unsupported chromosome requested
+  return (\@var_ann, $nextPageToken) unless -e $file;
+ 
+  my $parser = Bio::EnsEMBL::IO::Parser::VCF4Tabix->open( $file ) || die "Failed to get parser : $!\n";
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/ga4gh/variantannotations.pm]: This kind of initialisation should probably not be left until user request, unless you've got millions of file handles in play. Consider initialising as a Moose property with a builder sub{}

File: lib/EnsEMBL/REST/Model/ga4gh/variants.pm [hunk 5]
```diff
@@ -376,7 +382,41 @@ sub getSingleCallSets{
...
+  my ($chr, $pos) = split/\_/, $varname ;
+
+  $data->{start}         = $pos - 1;
+  $data->{end }          = $pos;
+  $data->{referenceName} = $chr;
+  $data->{variantSetId}  = $variantSetId; 
 
+  ## load VCFcollections object for variantSet 
+  $data->{vcf_collection} = $self->context->model('ga4gh::ga4gh_utils')->fetch_VCFcollection_by_id($data->{variantSetId});
+  $self->context()->go( 'ReturnError', 'custom', [ " Failed to find the specified variantSetId"])
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/ga4gh/variants.pm]: Don't do the controller's job in the model. Models are not supposed to trigger responses to users, they should only throw exceptions for the controller to handle, or else return some data.

**Comment by nerdstrike** [comment]
That's better.

**Comment by magaliruffier** [comment]
some failures on the gafeatures test now, it seems to be missing some expected childIds

**Comment by sarahhunt** [comment]
Thanks @magaliruffier ! I've pushed the update..

**Comment by magaliruffier** [comment]
it's looking fine now, thanks for the update

**Pull request merged**